store.go

```diff
@@ -826,36 +838,90 @@ func (s *store) GraphDriver() (drivers.Driver, error) {
 // LayerStore obtains and returns a handle to the writeable layer store object
 // used by the Store.  Accessing this store directly will bypass locking and
 // synchronization, so it is not a part of the exported Store interface.
+// Deprecated: Use RWLayerStore() instead
 func (s *store) LayerStore() (LayerStore, error) {
-	s.graphLock.Lock()
-	defer s.graphLock.Unlock()
-	if s.graphLock.TouchedSince(s.lastLoaded) {
-		s.graphDriver = nil
-		s.layerStore = nil
-		s.lastLoaded = time.Now()
-	}
+	return s.RWLayerStore()
+}
+
+// RWLayerStore obtains and returns a handle to the writeable layer store object
+// used by the Store.  Accessing this store directly will bypass locking and
+// synchronization, so it is not a part of the exported Store interface.
+func (s *store) RWLayerStore() (LayerStore, error) {
+	s.resetIfNeeded()
+
+	// Memoized read-write layer store
 	if s.layerStore != nil {
 		return s.layerStore, nil
 	}
-	driver, err := s.getGraphDriver()
+
+	rlpath, glpath, driver, err := s.getLayerPaths()
 	if err != nil {
 		return nil, err
 	}
-	driverPrefix := s.graphDriverName + "-"
-	rlpath := filepath.Join(s.runRoot, driverPrefix+"layers")
-	if err := os.MkdirAll(rlpath, 0700); err != nil {
+
+	rls, err := s.newLayerStore(rlpath, glpath, driver, true)
+	if err != nil {
 		return nil, err
 	}
-	glpath := filepath.Join(s.graphRoot, driverPrefix+"layers")
-	if err := os.MkdirAll(glpath, 0700); err != nil {
+
+	s.layerStore = rls
+	return s.layerStore, nil
+}
+
+// ROLayerStore obtains and returns a handle to the read-only layer store object
+// used by the Store.
+func (s *store) ROLayerStore() (ROLayerStore, error) {
+	s.resetIfNeeded()
+
+	// Memoized read-only layer store
+	if s.roLayerStore != nil {
+		return s.roLayerStore, nil
+	}
+
+	rlpath, glpath, driver, err := s.getLayerPaths()
+	if err != nil {
 		return nil, err
 	}
-	rls, err := s.newLayerStore(rlpath, glpath, driver)
+
+	rls, err := s.newLayerStore(rlpath, glpath, driver, false)
 	if err != nil {
 		return nil, err
 	}
-	s.layerStore = rls
-	return s.layerStore, nil
+
+	s.roLayerStore = rls
+	return s.roLayerStore, nil
+}
+
+// getLayerPaths retrieves the run directory (`rlpath`) the layer directory
+// (`glpath`) and the currently configured storage driver for the store
+func (s *store) getLayerPaths() (rlpath, glpath string, driver drivers.Driver, err error) {
+	driver, err = s.getGraphDriver()
+	if err != nil {
+		return "", "", nil, err
+	}
+	driverPrefix := s.graphDriverName + "-"
```

Comment:
I'd say have
```suggestion
layers := s.graphDriverName + "-" + "layers"
```
and use it below.

Reply:
Fixed as suggested